components/server/src/bitbucket-server/bitbucket-server-repository-provider.ts

```diff
@@ -100,4 +96,13 @@ export class BitbucketServerRepositoryProvider implements RepositoryProvider {
         //     authorAvatarUrl: commit.author?.user?.links?.avatar?.href,
         // };
     }
+
+    getUserRepos(user: User): Promise<string[]> {
+        throw new Error("getUserRepos not implemented.");
+    }
+
+    hasReadAccess(user: User, owner: string, repo: string): Promise<boolean> {
+        throw new Error("Method not implemented.");
```

Comment:
in that case `return true` would be a better default value, I guess. even if not correct in all cases, it should be stopper in general.

Reply:
Good point, thanks @AlexTugarev. But `return true` would mean "allow anyone to open snapshots of Bitbucket server repositories that are potentially private", which goes somewhat against the intent of this change (i.e. restrict snapshot access to only users who are allowed to read the repository).

Then again, if you have an account for a particular Bitbucket server, maybe it means you should have access to all repos/snapshots? (E.g. maybe these Bitbucket servers are for companies, and only employees can have an account there?)

Reply:
Do I read the code right, so that I see no way to open a workspace any more for BB Server if the `Method not implemented` error remains in place? If so, please change to the default `true`, and add a `TODO` in there. We're not expecting to make it feature complete very soon.

BB Server support is an alpha feature for Gitpod Self-Hosted only.

Also, the nature of the calling site is to check for a relevant git connection first, so this is not big issue. 

> goes somewhat against the intent of this Pull Request

I agree and understand. Let's call it an exception and maybe put a short `TODO` comment in there. That BB Server support is alpha, so ... 🤷🏻‍♂️

Reply:
> Do I read the code right, so that I see no way to open a workspace any more for BB Server if the `Method not implemented` error remains in place?

Only when you're trying to open a Snapshot made on a Bitbucket server repo.

Creating regular workspaces or prebuilds for Bitbucket server repos continues to work as before.

However, now that you mention it, `hasReadAccess` is also called when trying to watch headless logs. 🤔 I wonder what happens in that case. Maybe returning `false` is safer here.

Reply:
Or, actually, returning `true` and calling that an exception as you suggested.

It's a small security concern, but going from:
- snapshots can be opened by anyone

to:
- snapshots are restricted to repo readers, except for Bitbucket servers where they're still unrestricted

is already a good first step I think.